internal/dispatch/singleflight/singleflight.go

```diff
@@ -53,19 +55,25 @@ func (d *Dispatcher) DispatchCheck(ctx context.Context, req *v1.DispatchCheckReq
 
 	keyString := hex.EncodeToString(key)
 
-	// Check if the key has already been part of a dispatch, for the *same* request ID. If so, this represents a
+	// Check if the key has already been part of a dispatch. If so, this represents a
 	// likely recursive call, so we dispatch it to the delegate to avoid the singleflight from blocking it.
-	requestID := req.Metadata.RequestId
-	existed := d.checkByDispatchKey.Add(keyString, requestID)
-	defer d.checkByDispatchKey.Remove(keyString, requestID)
-
-	if existed {
-		// Likely a recursive call.
+	// If the bloom filter presents a false positive, a dispatch will happen, which is a small inefficiency
+	// traded-off to prevent a recursive-call deadlock
+	serializedBloom, err := validateTraversalRecursion(keyString, req.Metadata)
+	if errors.Is(err, ErrLoopDetect) {
+		singleFlightCount.WithLabelValues("DispatchCheck", "loop").Inc()
 		return d.delegate.DispatchCheck(ctx, req)
 	}
 
+	if err != nil {
+		return &v1.DispatchCheckResponse{Metadata: &v1.ResponseMeta{DispatchCount: 1}}, err
+	}
+
+	clonedReq := req.CloneVT()
```

Comment:
This is going to be very inefficient - we should endeavor not to clone the response

Reply:
I can address this but I have concerns about it. The moment the same dispatch request is reused by two dispatch calls, it'd be leaking the state of the bloom filter. The worst that can happen is that singleflight will by bypassed, which is not terrible, mind you, but it would exhibit that behaviour and be difficult to understand why.